x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -223,14 +224,42 @@ public Map<String, Object> getMetadata() {
     public Authentication maybeRewriteForOlderVersion(Version olderVersion) {
         // TODO how can this not be true
         // assert olderVersion.onOrBefore(getVersion());
-        Authentication newAuthentication = new Authentication(
-            getUser(),
-            maybeRewriteRealmRef(olderVersion, getAuthenticatedBy()),
-            maybeRewriteRealmRef(olderVersion, getLookedUpBy()),
-            olderVersion,
-            getAuthenticationType(),
-            maybeRewriteMetadataForApiKeyRoleDescriptors(olderVersion, this)
-        );
+
+        final Map<String, Object> newMetadata = maybeRewriteMetadataForApiKeyRoleDescriptors(olderVersion, this);
+
+        final Authentication newAuthentication;
+        if (isRunAs()) {
+            // The lookup user for run-as currently doesn't have authentication metadata associated with them because
+            // lookupUser only returns the User object. The lookup user for authorization delegation does have
+            // authentication metadata, but the realm does not expose this difference between authenticatingUser and
+            // delegateUser so effectively this is handled together with the authenticatingSubject not effectiveSubject.
+            newAuthentication = new Authentication(
+                new Subject(
+                    effectiveSubject.getUser(),
+                    maybeRewriteRealmRef(olderVersion, effectiveSubject.getRealm()),
+                    olderVersion,
+                    effectiveSubject.getMetadata()
+                ),
+                new Subject(
+                    authenticatingSubject.getUser(),
+                    maybeRewriteRealmRef(olderVersion, authenticatingSubject.getRealm()),
+                    olderVersion,
+                    newMetadata
+                ),
+                type
+            );
+        } else {
+            newAuthentication = new Authentication(
+                new Subject(
+                    authenticatingSubject.getUser(),
+                    maybeRewriteRealmRef(olderVersion, authenticatingSubject.getRealm()),
+                    olderVersion,
+                    newMetadata
+                ),
+                type
+            );
```

Comment:
This is the only place that is more bloated after refactoring. In all other places, the change is a simplification.